08-rails/indirect-relations.md

```diff
@@ -0,0 +1,127 @@
+
+A common scenario when working with databases is to have two models that are related indirectly, through a third model. For example in our library application `Author` and `Genre` are related through the `Book` model.
+
+![Author and Genre are related through Book](images/indirect-model-relation.png)
+<!-- https://www.draw.io/#G1n2OylcmI2jEzvX_W76-0OpfHfk5QPnOE -->
+
+**Questions:** Imagine that on the details page for an author, we wanted to list all the genres they have written. The list of genres should be de-duplicated (each genre should not appear more than once).
+- What would the code look like to build this list of genres?
+- Where should this code live?
+- What view code would you need to display this list?
+
+![Genre list for an author](images/indirect-relation-ui.png)
+
+There are two pieces of work here. Generating the list of genres is business logic, and belongs in the model:
+
+```ruby
+# app/models/author.rb
+class Author < ApplicationRecord
+  # ...
+  def genres
+    genres = []
+    books.each do |book|
+      genres += book.genres
+    end
+    return genres
+  end
+end
+```
+
+This method gives us an array of `genre`s, as we would expect from a model method. We now need to write some display logic to turn that list into some HTML. Here is one way to approach it:
+
+```html
+<!-- app/views/authors/show.html.erb -->
+<p>
+  <strong>Genres:</strong>
+  <%= @author.genres.uniq.map { |g| g.name.capitalize }.join(', ') %>
+</p>
+```
+
+## Rails Syntax for Indirect Relations
+
+Turns out this kind of problem is very common when working with databases. We might phrase it generically as "find all the records that are two steps away". For example, in our ride share application we might want to find a list of all passengers who have ridden with a certain driver.
+
+Because it's so common, Rails gives us a way to come up with this list automatically using the `through` keyword argument to the `has_many` helper method:
+
+```ruby
+# app/models/author.rb
+class Author < ApplicationRecord
+  has_many :books
+  has_many :genres, through: :books
+
+  # ...
+  # remove the genres method we defined before
+  # ...
+end
+```
+
+There are two parts to this line of code. `has_many :genres` specifies both what table we're building a relation to, and what the method we're generating will be called. The `through: :books` tells us how to get there: follow the `books` relation for this author (defined on the line above).
+
+If we refresh the page, we should see exactly the same behavior. All we've done is remove some boilerplate.
+
+Note that we did **not** need to create a new migration for this work. We've upgraded the ActiveRecord model, but have not changed our database schema.
+
+**Question:** What would we do if we needed to get the list of authors for a given genre?
+
+### Singular Indirect Relations
+
+<!-- QUESTION: do we need this section? Lesson would be more focused without it. -->
```

Comment:
I think it’s nice to know but not necessary as part of this discussion. Since there are already two important new relationships, I’d lean towards leaving this off and linking a resource to it below.

Reply:
I agree with Kari's thought-- If I were giving this lecture as-is, I would de-emphasize this section. I wish we had the time to give it justice though!